### PR TITLE
Fixes #24416 - Add UI devel docs into storybook

### PR DESCRIPTION
**Pull request by tstrachota**
I'm adding some materials that I put together for the "UI in Foreman" session I did for our crew in Brno:
- adds support for displaying markdown with syntax highlighting in storybook
- moves existing component stories under "Components/"
- adds section "Introduction" with devel docs

I take this as an initial step. I expect that we extend the docs with additional info in future. We can also give it better structure and better formatting of the component stories later.

### CODE REVIEW

**Comment by tstrachota** [comment]
@ares @waldenraines @amirfefer @amirfefer could you please check for typos and factual accuracy?

**Comment by tstrachota** [comment]
Ooops, I used wrong ticket number, fixed now.

File: webpack/assets/javascripts/react_app/components/BreadcrumbBar/BreadcrumbBar.stories.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { boolean, number, withKnobs } from '@storybook/addon-knobs';
 import { action } from '@storybook/addon-actions';
 import BreadcrumbBar from './BreadcrumbBar';
 
-storiesOf('BreadcrumBar', module)
+storiesOf('Components/BreadcrumBar', module)
```
Comment (adamruzicka) [inline webpack/assets/javascripts/react_app/components/BreadcrumbBar/BreadcrumbBar.stories.js]: Could you fix the typo in the name of the component when you're touching this?

File: webpack/stories/docs/addingDependencies.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+npm install --save <pkg>@<version>
+npm install --save-dev <pkg>@<version>
+```
+That will automatically edit pacakge.json at correct places for you.
+
+### Packaging
+
+The Foreman distributes npm dependencies in separate rpm files. Therefore you have to make sure the new dependency is packaged as rpm. You can find the instructions about how to create an rpm from an npm packge in the [foreman-packaging](https://github.com/theforeman/foreman-packaging/tree/rpm/develop#adding-npm-packages) repo.
+
+**Watch out:** adding/updating npm dependencies is currently very frequent source of errors. There's a check on github that won't let you merge a PR that touches package.json unless the packaging team approves the change, but it's good to be careful in this area. More details in the following section.
```
Comment (adamruzicka) [inline webpack/stories/docs/addingDependencies.md]: We should probably use properly capitalized `GitHub` in docs

File: webpack/stories/docs/internationalization.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# Translations
+
+All strings need to be translated. We use gettext's underscore `__()` function as you know it from our asset pipeline code.
+
+Note that for the time being the underscore function is available in the global `window` object and translations are handled outside of our webpack-processed code and there's no need to import it. This is most likely subject to change in future.
+
+
+## Interpolation
+
+Make sure that the strings inside the underscore function doesn't contain any interpolated values.
```
Comment (adamruzicka) [inline webpack/stories/docs/internationalization.md]: "[the strings] don't contain"

**Review by adamruzicka — COMMENTED** [review]
I just quickly glanced over this and left some nitpicks inline

**Comment by tstrachota** [comment]
Thanks @adamruzicka, updated

File: webpack/stories/docs/addingNewComponent.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
+# Adding components
+
+## Where to put them
+
+Components are stored in `webpack/assets/javascripts/react_app/components/`. Each component should be placed in its own subfolder that respects the following structure:
+
+```
+─ components/<COMPONENT_NAME>/
```
Comment (amirfefer) [inline webpack/stories/docs/addingNewComponent.md:8]: Selectors are missing here

**Review by amirfefer — COMMENTED** [review]
Thanks @tstrachota , I like it! 
I think we should add selectors convention as well (#5778)

**Comment by lizagilman** [comment]
@tstrachota looks good!

File: webpack/stories/docs/addingNewComponent.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+```
+
+**Example:**
+```erb
+<%= mount_react_component('PowerStatus', '#power', {:id => host.id, :url => power_host_path(host.id)}.to_json) %>
+```
+
+## Before you start writing a new component
+
+It's worth checking pfatternfly-react [Github repository](https://github.com/patternfly/patternfly-react) and [storybook](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html) to make sure such component doesn't exist yet. Also consider if your component is universal enough to be used in other projects. In such case it might make sense to add it to patternfly-react instead.
```
Comment (lizagilman) [inline webpack/stories/docs/addingNewComponent.md]: patternfly-react (redundant 'f')

File: webpack/stories/docs/addingNewComponent.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+   ├─ components/                    ┈ folder for nested components if needed
+   ├─ __tests__/                     ┈ folder for tests
+   ╰─ index.js                       ┈ redux connected file
+```
+
+## Storybook
+
+[Storybook](https://storybook.js.org/) is an isolated environment for developing and demonstrating components. It serves as a nice documentation of a component usage.
+
+Each components that is supposed to be used at multiple places in the application should add its stories into the "Components" menu. The stories should be placed in a file `<COMPONENT_NAME>.stories.js` in the component's main folder. The storybook in Foreman loads all `*.story.js` files from `webpack/` automatically.
```
Comment (lizagilman) [inline webpack/stories/docs/addingNewComponent.md]: each component ? (singular)

File: webpack/stories/docs/addingNewComponent.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  - **render** - static rendering, more details in [docs](https://github.com/airbnb/enzyme/blob/master/docs/api/render.md)
+
+### Running the tests
+
+All tests can be executed with `npm test`.
+
+If you want to run only a single test use `jest` directly: `./node_modules/.bin/jest <path/to/some.test.js>`
+This is useful especially for debugging because it doesn't hide console output.
+
+Linter (code style checking) can be executed with `npm run lint`.
```
Comment (lizagilman) [inline webpack/stories/docs/addingNewComponent.md]: should we add here - 
a. a simple way to fix lint errors is with `--fix` 
b. a recommendation to use code formatter
?

**Comment by tstrachota** [comment]
Thanks for your reviews, updated.

File: webpack/stories/docs/addingNewComponent.md (hunk 4 above)
Comment (tstrachota) [reply webpack/stories/docs/addingNewComponent.md:8]: I added `<COMPONENT_NAME>Selectors.js` down below. Selectors probably deserve their own paragraph, but I prefer adding it outside of this PR.

Comment (boaz0) [reply webpack/stories/docs/addingNewComponent.md:8]: @tstrachota I am also missing `<COMPONENT_NAME>Consts.js` where constants such as action-types are stored.

Comment (tstrachota) [reply webpack/stories/docs/addingNewComponent.md:8]: Good point @ripcurld, added

**Comment by tstrachota** [comment]
Any more comments? @ohadlevy would you mind to review and possibly merge?

**Comment by ohadlevy** [comment]
@tstrachota mind rebasing and checking the tests? thanks.

File: package.json [hunk 8]
```diff
@@ -36,12 +36,15 @@
     "expose-loader": "~0.6.0",
     "extract-text-webpack-plugin": "^3.0.0",
     "file-loader": "^0.9.0",
+    "highlight.js": "^9.12.0",
```
Comment (ohadlevy) [inline package.json:39]: if those depenecides are used only for storybook, we should add them to the build ignore white list, similar to https://github.com/theforeman/foreman-packaging/pull/2846

File: webpack/stories/docs/addingDependencies.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
+
+# Adding or updating NPM dependencies
+
+The easiest way of adding dependencies into [package.json](https://github.com/theforeman/foreman/blob/develop/package.json) is executing `npm install` with `--save` or `--save-dev` for devel dependencies:
+```
+npm install --save <pkg>@<version>
+npm install --save-dev <pkg>@<version>
+```
```
Comment (ohadlevy) [inline webpack/stories/docs/addingDependencies.md:8]: might add --no-optional to reduce node modules size (e.g. if you dont use phatomjs)

**Review by ohadlevy — CHANGES_REQUESTED** [review]
great work, thanks @tstrachota 

Added a few comments, plus, would mind adding a section about ```npm run analyze``` ?

File: webpack/stories/docs/gettingStarted.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+```
+─ webpack/                 ┈ all webpack processed code
+   │
+   ├─ stories/             ┈ storybook config (this pages)
+   ╰─ assets/javascripts/  ┈ es6 code for erb pages, some still contain jQuery
+       │
+       ╰─ react_app/       ┈ react components and related code
+```
+
+More detailed description of a folder structure for components is in chapter [Adding new component](TODO).
```
Comment (ohadlevy) [inline webpack/stories/docs/gettingStarted.md]: its no longer todo right? :)

File: package.json (hunk 8 above)
Comment (tstrachota) [reply package.json:39]: Done, see https://github.com/theforeman/foreman-packaging/pull/2851 and https://github.com/theforeman/foreman-packaging/pull/2852

File: webpack/stories/docs/addingDependencies.md (hunk 9 above)
Comment (tstrachota) [reply webpack/stories/docs/addingDependencies.md:8]: I added a mention here: https://github.com/theforeman/foreman/pull/5883/files#diff-de4d6b944c9164c2bc2c69ecdf26d018R25

**Comment by tstrachota** [comment]
Thanks @ohadlevy, updated

> would mind adding a section about npm run analyze ?

I prefer adding it in a separate PR. There's plenty of more docs in my mind that we could add. I'd like to merge what we have now and iterate on it later.

**Review by ohadlevy — APPROVED** [review]

**Comment by ohadlevy** [comment]
[test foreman]

**Comment by ohadlevy** [comment]
It seems that Travis tests are failing consistently, can you please have a look?

**Review by ohadlevy — APPROVED** [review]

**Comment by ohadlevy** [comment]
thanks @tstrachota, it would be great if you could add a task that auto deploy the storybook each time a relevant commit is merged.

**Comment by tstrachota** [comment]
Thanks @ohadlevy, that makes sense. I'll try to take a look at it.

**Comment by ohadlevy** [comment]
I've published this manually for now at https://theforeman.github.io/foreman

**Comment by ekohl** [comment]
https://theforeman.github.io/foreman/?selectedKind=Components%2FFactChartModal&selectedStory=ModalOpen&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel shows `n__ is not defined` for me.

**Comment by ohadlevy** [comment]
this is most likely a bug in the stroybook, please open a bug on it (I
assume the reason is a refactoring that did not take into account the
storybook code - adding a test would be nice).

**Pull request merged**